### PR TITLE
Add array/domain creation perftest

### LINKED ISSUE #12930: Create perf tests to track distributed array creation
Create some comm-count and performance tests to track the amount of work that goes into setting up a distributed array. Today it's quite expensive and not an area we've spent much time optimizing with the thought that most distributed arrays will be long lived (and as a result we've spent our time optimizing operations on the arrays, not the creation itself.)

In some cases we've seen it's slower than we want (https://github.com/chapel-lang/chapel/issues/12926), and we have some UPC/SHMEM users who intuitively expect it to be less expensive than it is. We should improve our distributed array creation time, but before getting to that we should start tracking performance.

I think Block, Stencil, and Cyclic are the most important ones currently, but there may be others that are worth tracking. It's probably also worth having 2 variants for small (1M element) and large (1/4 mem) arrays.

__edit__:

 I don't see different comm counts between small and large arrays. 

It might be interesting to have comm counts for the array creation helpers like `newCyclicArr()` since they currently current a new dist domain per call.

### PR DESCRIPTION
**Pull request by e-kayrakli**
Adds a performance test to keep track of distributed array/domain creation times.

Tests the following:

- Domain init and deinit
- 1 element-per-locale array init and deinit
- 1M element array init and deinit
- 1/4 of total memory array init and deinit

All tests measure time for Block, Cyclic and BlockCyc

I tested the added test with the following on XC:
`start_test --perflabel ml-`
`start_test`

Also tested the correctness mode on chapcs with standard and gasnet configs

### CODE REVIEW

File: test/performance/array/distCreate-large.ml-perf.graph [hunk 1]
```diff
@@ -0,0 +1,5 @@
+perfkeys: blockDom:, blockArr:, cyclicArr:, cyclicDom:, blockCyclicDom:, blockCyclicArr:
```
Comment (ronawho) [inline test/performance/array/distCreate-large.ml-perf.graph]: nit, but can you swap the  `cyclicArr` and `cyclicDom` order so it's consistent with the others?

File: test/performance/array/distCreate-large.ml-perf.graph [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+files: distCreate-large 
+ylabel: Time (seconds)
+graphname: distCreate-small
```
Comment (ronawho) [inline test/performance/array/distCreate-large.ml-perf.graph]: I don't think we use/need graphname anymore  (I'd look at some of the other `ml-perf.graph` files as a template)

File: test/performance/array/distCreate-large.ml-perf.graph [hunk 3]
```diff
@@ -0,0 +1,5 @@
+perfkeys: blockDom:, blockArr:, cyclicArr:, cyclicDom:, blockCyclicDom:, blockCyclicArr:
+files: distCreate-large
```
Comment (ronawho) [inline test/performance/array/distCreate-large.ml-perf.graph]: `files` requires a list of files (that has the same length as `perfkeys`) If you want to use the same file for all `perfkeys` it's `repeat-files`

File: test/performance/array/distCreate.chpl [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+type elemType = int;
+const totalMem = + reduce Locales.physicalMemory();
+
+config param correctness = true;
+config const nElemsSmall = if correctness then 100 else 1000000;
+config const memRatio = .25;
+config const nElemsLarge = ((totalMem*memRatio)/numBytes(elemType)):int;
+config const small = true;
+
+const nElems = if small then nElemsSmall else nElemsLarge;
```
Comment (ronawho) [inline test/performance/array/distCreate.chpl]: Not sure it matters, but I'd take a look at `test/scan/scanPerf.chpl`. I try to copy problem size computations from existing tests, but I don't mind keeping this as-is.

File: test/performance/array/distCreate-small.ml-perf.graph [hunk 5]
```diff
@@ -0,0 +1,5 @@
+perfkeys: blockDom:, blockArr:, cyclicArr:, cyclicDom:, blockCyclicDom:, blockCyclicArr:
+files: distCreate-small 
+ylabel: Time (seconds)
+graphname: distCreate-small
```
Comment (ronawho) [inline test/performance/array/distCreate-small.ml-perf.graph]: `cyclicArr` / `cyclicDom` order swap, use `repease-files` and drop` graphname`

File: test/performance/array/distCreate.chpl [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+{
+  startDiag();
+  const blockDom = localDom dmapped Block(boundingBox=localDom);
+  endDiag("blockDom", blockDom);
+
+  startDiag();
+  const blockArr: [blockDom] elemType;
+  endDiag("blockArr", blockArr);
+}
```
Comment (ronawho) [inline test/performance/array/distCreate.chpl]: I wonder what we want to do about tracking array destruction time

File: test/performance/array/distCreate.ml-compopts [hunk 7]
```diff
@@ -0,0 +1 @@
+-scorrectness=false
```
Comment (ronawho) [inline test/performance/array/distCreate.ml-compopts]: Does correctness need to be a param? Might be a little cleaner as a const, and then throw this in the ml-execopts

Comment (ronawho) [reply test/performance/array/distCreate.ml-compopts]: This is also a nit, but I tend to have the default mode be the one I'm most interested in if I'm running this test manually i.e. for things designed as performance tests I tend to have the performance mode be the default (though I'm not 100% consistent with that.)

File: test/performance/array/distCreate.ml-keys [hunk 8]
```diff
@@ -0,0 +1,6 @@
+blockDom:
+blockArr:
+cyclicArr:
+cyclicDom:
```
Comment (ronawho) [inline test/performance/array/distCreate.ml-keys]: Swan cyclicDom/cyclicArr order

File: test/performance/array/distCreate.skipif [hunk 9]
```diff
@@ -0,0 +1 @@
+CHPL_COMM==none
```
Comment (ronawho) [inline test/performance/array/distCreate.skipif]: Personally I don't see any reason to skip this for comm=none

File: test/performance/array/distCreate-large.ml-perf.graph (hunk 2 above)
Comment (ronawho) [reply test/performance/array/distCreate-large.ml-perf.graph]: Oh, this also needs to be `distCreate-small.dat` (For some reason we require the .dat extension here, but not in compopts/execopts ) -- FWIW I usually don't remember this off the top of my head. I normally copy an existing test as a template, and I also look at the graph to make sure it looks right

**Comment by ronawho** [comment]
Left a bunch of code comments but at a higher level -- I wonder what all we want to track here. I wouldn't expect domain creation to change with the size of the array, so I wondering if it makes sense to split out the array and domain graphs. I also think it'd be worthwhile to track another array size where you basically get 1 element per locale (which we have seen in user code.) Left to my own devices I'd probably have 4 graphs:
  - domain creation time
  - single element per locale array creation time 
  - 1M elements array creation time
  - 1/4 memory array creation time

I'm also thinking we may want multiple trials since all of these operations are well under a second. We could either do a fixed number of trials, or we could make this a rate graph and track how many domains/array we can create per second or something.

I'd also like to think about what we want to do for tracking domain/array destruction time.

**Comment by e-kayrakli** [comment]
@ronawho thanks for the review. All the minor ones make sense. I addressed them and confirmed that the graph is produced already. For the high-level questions you bring up:

- Having those 4 graphs/tests makes sense to me. I will change the PR to do that.
- Measuring time of multiple trials also makes sense to me. I am not a big fan of measuring throughput of array creation. I'd find time to create an array much more relatable to real-world scenarios. Would it make sense to start the testing with time measurement and if we observe too much noise, we switch to throughput?
- I don't think there is a clear way to measure the destruction time, right? I am wondering if computing it ourselves is a reliable method? Within a test we can create an array/domain and measure creation time. Then declare an identical array/domain in a tight scope and measure its performance, and get the difference. Not ideal, but maybe the only neat way? 
   - We can also consider reporting the time to create+destroy an array. Because doing one right  after the other might have better cache locality and give us readings that are potentially better than what would happen in a real-world scenario.
   - Another alternative I can think of to avoid the cache effect is to declare objects and put them in a large array (maybe larger than LLC) and measure the overall destruction time, reporting the time divided by number of arrays. (assuming the main array is local and takes little time to destruct, and the noise it creates is acceptable)

**Comment by ronawho** [comment]
> Measuring time of multiple trials also makes sense to me. I am not a big fan of measuring throughput of array creation. I'd find time to create an array much more relatable to real-world scenarios. Would it make sense to start the testing with time measurement and if we observe too much noise, we switch to throughput?

Makes sense to me

> I don't think there is a clear way to measure the destruction time, right? 

I was thinking something like:

```chpl
var init, destroy: Timer;
{
  init.start();
  var A: [D] int;
  init.stop();
  destroy.start();
}
destroy.stop();
```


As far as cache -- I'm not too worried about getting timings that are better than they might be in real-world usage because stuff is fresh in the cache. I think this test will mostly be to track relative performance (have things gotten better/worse) rather than the absolute time.

**Comment by e-kayrakli** [comment]
Among other things, the last commits made this PR adding 8 graphs:

- Domain init and deinit
- 1 element-per-locale array init and deinit
- 1M element array init and deinit
- 1/4 of total memory array init and deinit

All tests measure time for Block, Cyclic and BlockCyc

**Pull request merged**